### PR TITLE
doc: add OTA section to device management

### LINKED ISSUE #42240: doc: Propose adding a section on Over-the-Air (OTA) updates
**Is your enhancement proposal related to a problem? Please describe.**
I was looking to reference Zephyr docs on over-the-air updates but found nothing on the topic.

**Describe the solution you'd like**
Add an OTA page and making it a bullet point in the [Device Management](https://docs.zephyrproject.org/latest/guides/device_mgmt/index.html) section (plus tidying up that page).

This page already links to the [DFU page](https://docs.zephyrproject.org/latest/guides/device_mgmt/dfu.html) which discusses MCUboot. We are using the MCUboot bootloader to flash binaries OTA. It makes sense to discuss OTA as part of device management.

**Describe alternatives you've considered**
One alternative would be to add an Over-the-Air heading to the end of the DFU page. However, that page mentions not discussing "transport or management protocols" and links back to the device management page. As OTA will certainly gain popularity, it's better to  have a dedicated page that can grow.

**Additional context**
I work at Golioth and we plan to add documentation to this section in the future.

### PR DESCRIPTION
**Pull request by szczys**
This adds an OTA page to the device management section of the docs and links to some examples.

Zephyr docs should mention OTA:
- Define Over-the-Air update
- Indicate OTA can be used with MCUboot
- Link to Golioth sample code and documentation of a working example of
OTA update process
- Closes #42240

Signed-off-by: Mike Szczys <mike@golioth.io>

### CODE REVIEW

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:12]: This doesn't parse. The devices aren't referred to by "OTA", the updates are. How about this instead:

```suggestion
connection, updates received over a wired connection (such as Ethernet)
are still commonly referred to as OTA updates. This approach requires server
```

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:55]: IMO this is **way** too fluffy for a technical page; was it copy/pasted from some marketing copy for UpdateHub? I think that's not appropriate if so. Let's focus on the technical aspects of the solution here, please.

Also, why do we care about updating Linux-based devices here? This is Zephyr, not Linux.

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:27]: What does it mean to "include Zephyr OTA updates"?

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:34]: Security is mentioned here but not anywhere else. For parallelism, I think it makes sense to:

- move the blanket statement that security is a concern higher up, outside of the sections for any individual solution
- document the security related bits of the implementation for each solution

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:58]: Were lwm2m firmware updates via the standard Firmware object (see `CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT`) intentionally omitted? If so, why?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Sounds like a good idea to me! A few concerns about the specifics.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:12]: Good feedback, thanks! I'll incorporate this change in an update to the commit.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:55]: Good criticism, and yes I just pulled the paragraph from their web page since I'm not familiar with the project. I have reworked this entry following your feedback.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:27]: Heh, fair enough. I'm removing Zephyr from that sentence.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:34]: Will do.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:58]: It was omitted because I didn't know about it. It seems the firmware upgrade support in LwM2M is largely undocumented (in the Zephyr docs) and not included in the LwM2M client sample. I'm adding a section to reference it as an option.

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:44]:
```suggestion
is downloaded and installed. MCUboot can be used to verify the signature before flashing the firmware.
```

I think it's worth making some change along these lines because hawkbit itself doesn't require mcuboot; that's an implementation decision made by the sample. Feel free to reword if you want.

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:55]: Similarly, does updatehub require mcuboot, or is that an implementation decision in upstream zephyr's updatehub integration?

File: doc/guides/device_mgmt/ota.rst
Comment (mbolivar-nordic) [inline doc/guides/device_mgmt/ota.rst:74]:
```suggestion
:kconfig:`CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT`. Devices securely connect to an
```

When referring to Kconfig options within the Zephyr documentation, it's preferred to use the `:kconfig:` Sphinx role like the above. This generates a link to a page with more information about the option, such as its dependencies, help text, etc.

(Similarly, when referring to devicetree compatibles, use the `:dtcompatible:` role to get a link into the appropriate bindings index page.)

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Thank you for the updates and addressing my feedback! I've got a few follow-up comments related to the new changes, but after that I think I'm +1.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:74]: I'm glad you mentioned this. When I added this section I first tried using the `:kconfig:` identifier. It generates the link but the url looks like it doesn't go where you think it should:
```
~/golioth-compile/zephyr/doc/_build/html/reference/kconfig/dummy-syms.html#std-kconfig-CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT
```

I'll put it in and submit an updated commit. Thanks!

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:44]: Good suggestion, thanks.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:55]: I don't know the answer to this; the sample uses MCUboot. I think your wording from the previous comment also makes sense here.

File: doc/guides/device_mgmt/ota.rst
Comment (utzig) [inline doc/guides/device_mgmt/ota.rst:20]: I find this terminology a bit weird for MCUboot users, it would be better to say something like ", then upgraded to using the ..."

Comment (utzig) [reply doc/guides/device_mgmt/ota.rst:20]: Also instead of pointing to the MCUboot website, you could just refer to the Zephyr docs which already have some MCUboot specific information with: ref:\`mcuboot`

File: doc/guides/device_mgmt/ota.rst
Comment (utzig) [inline doc/guides/device_mgmt/ota.rst:32]: "before the upgrade occurs", and so on...

File: doc/guides/device_mgmt/ota.rst
Comment (utzig) [inline doc/guides/device_mgmt/ota.rst:19]: maybe "... into an unoccupied code partition, usually named `slot1_partition`, then upgraded using..." ?

**Review by utzig — COMMENTED** [review]
Looks generally good, I left a few comments.

**Review by mbolivar-nordic — COMMENTED** [review]
LGTM assuming @utzig's comments are addressed too

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:20]: > ", then upgraded to using the ..."

Got it, thanks!

> Also instead of pointing to the MCUboot website, you could just refer to the Zephyr docs which already have some MCUboot specific information with: ref:`mcuboot`

I looked and I don't see a Zephyr docs page on MCUboot. Am I just missing it? The DFU page which I've already linked in this paragraph discuses it.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:32]: Got it, thanks!

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst:19]: Sounds good, thanks!

File: doc/guides/device_mgmt/ota.rst
Comment (utzig) [reply doc/guides/device_mgmt/ota.rst]: The MCUboot reference I mentioned is actually a section inside the DFU page, you can link to it using ":ref:\`mcuboot`"

**Comment by utzig** [comment]
@szczys I would suggest you don't mark issues as resolved, and instead leave it for maintainers.

**Comment by szczys** [comment]
@utzig my apologies. Thanks for letting me know, I will not do that in the future.

File: doc/guides/device_mgmt/ota.rst
Comment (szczys) [reply doc/guides/device_mgmt/ota.rst]: Thanks for the clarification, I've pushed this change.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by utzig — APPROVED** [review]
LGTM

**Review by ycsin — APPROVED** [review]

**Pull request merged**
